### PR TITLE
Fix `MetricCollection` dict keys when used with `ClasswiseWrapper` [wip]

### PR DESCRIPTION
**Pull request by sup3rgiu**
## What does this PR do?

When a `ClasswiseWrapper` is inserted into a `MetricCollection`, the resulting output dictionary does not take into account the specified name.

```py
import torch
from torchmetrics import MetricCollection
from torchmetrics.wrappers import ClasswiseWrapper
from torchmetrics.classification import MulticlassAccuracy, MulticlassRecall
labels = ["horse", "fish", "dog"]
metric = MetricCollection(
    {'my/name/': ClasswiseWrapper(MulticlassAccuracy(num_classes=3, average=None), labels, prefix='accuracy_'),
    'my/other_name/': ClasswiseWrapper(MulticlassRecall(num_classes=3, average=None), labels, prefix='recall_')}
)
preds = torch.randn(10, 3).softmax(dim=-1)
target = torch.randint(3, (10,))
metric(preds, target)  
```

The previous code produces the following output:
```
{'accuracy_horse': tensor(0.5000),
 'accuracy_fish': tensor(0.2500),
 'accuracy_dog': tensor(0.5000),
 'recall_horse': tensor(0.5000),
 'recall_fish': tensor(0.2500),
 'recall_dog': tensor(0.5000)}
```

While with proposed fix, the output is the following:
```
{'my/name/accuracy_horse': tensor(0.5000),
 'my/name/accuracy_fish': tensor(0.2500),
 'my/name/accuracy_dog': tensor(0.5000),
 'my/other_name/recall_horse': tensor(0.5000),
 'my/other_name/recall_fish': tensor(0.2500),
 'my/other_name/recall_dog': tensor(0.5000)}
```

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2361.org.readthedocs.build/en/2361/

### CODE REVIEW

**Comment by Borda** [comment]
@sup3rgiu please see the failing doctest, seems that chnege does not work as intended yet...

**Comment by Borda** [comment]
@sup3rgiu let's turn it to draft until failing tests are addressed 🦩

**Comment by Borda** [comment]
@SkafteNicki, what is missing here to make it land? :chipmunk:

**Comment by Borda** [comment]
@SkafteNicki, what is missing here to make it land? :chipmunk:

**Comment by Borda** [comment]
@sup3rgiu could you pls add some tests?

**Comment by SkafteNicki** [comment]
I finally looked what could be done regarding this PR and the sad reality is that doing this particular change would have have large consequences for how the output in general for metric collection is formatted, which is not really backwards compatible. This comes down to it also influencing nesting of other metric wrappers/classes. In addition, the `prefix` argument and the key of a classwise metric collection at this point serves somewhat the same purpose in naming the metric. So if we had an example like this (take from unittests):
```python
    m = MetricCollection(
        {
            "accuracy": ClasswiseWrapper(MulticlassAccuracy(num_classes=3, average=None), prefix="accuracy"),
            "recall": ClasswiseWrapper(MulticlassRecall(num_classes=3, average=None), prefix="recall"),
            "precision": ClasswiseWrapper(MulticlassPrecision(num_classes=3, average=None), prefix="precision"),
        },
        compute_groups=[["accuracy", "recall", "precision"]],
    )
```
the output dictionary with the proposed change would be `{'accuracyaccuracy0': tensor(0.), 'accuracyaccuracy1': tensor(0.2500), 'accuracyaccuracy2': tensor(0.), 'precisionprecision0': tensor(0.), 'precisionprecision1': tensor(0.2500), 'precisionprecision2': tensor(0.), 'recallrecall0': tensor(0.), 'recallrecall1': tensor(0.2500), 'recallrecall2': tensor(0.)}`.

Thus, the result that is wanted in this case can be achieved by adding the dict keys to the `prefix` argument
```python
import torch
from torchmetrics import MetricCollection
from torchmetrics.wrappers import ClasswiseWrapper
from torchmetrics.classification import MulticlassAccuracy, MulticlassRecall
labels = ["horse", "fish", "dog"]
metric = MetricCollection(
    {'my/name/': ClasswiseWrapper(MulticlassAccuracy(num_classes=3, average=None), labels, prefix='my/name/accuracy_'),
    'my/other_name/': ClasswiseWrapper(MulticlassRecall(num_classes=3, average=None), labels, prefix='my/other_name/recall_')}
)
preds = torch.randn(10, 3).softmax(dim=-1)
target = torch.randint(3, (10,))
metric(preds, target)
```
will return the wanted result: ` {'my/name/accuracy_horse': tensor(0.3333), 'my/name/accuracy_fish': tensor(0.5000), 'my/name/accuracy_dog': tensor(0.3333), 'my/other_name/recall_horse': tensor(0.3333), 'my/other_name/recall_fish': tensor(0.5000), 'my/other_name/recall_dog': tensor(0.3333)} `. I know this is a bit ugly solution, but I still think it is overall better than breaking backwards compatibility.

I am closing this PR.
